Tweak the tag API a bit.

samoht opened the pull request:
Simplify the way sources are tagged.

hannesm commented:
I like that this removes lines of code I didn't understand why they're around :)
is `tags` the right mechanism where (a) a source provides their (constant run-time) information, such as "unit of measurement", "instrumented function name (in the case of a timer)", but also external information "this is on VM host named 'foobar'".  for the latter I suspect I need a global mutable "tags to add to each measurement" (or a dynamically-bound variable - similar to the tags in Logs)?

samoht commented:
I am not sure to understand what you mean. 

The current design of `Metrics` has these 2 concepts:

- metrics *names*: a name is a family of graph. Name could be used to enable/disable the collection of metrics (TODO).
- metrics *tags*: a list of (name * type * value), with a fixed order, fixed types, fixed length only the values are varying. Tags are used to 1/ enable/disable the collection of metrics and 2/ index graphs: in the gnuplot backend every '"line" corresponds to a metrics name + tags. 
- metrics *data*: a list of (name * type * value * unit), and a timestamp. The API does not force the data to have a static shape, but it's better if it is (the gunplot backend will probably be confused currently, but that's fixable). Data are the actual stuff being plotted on the graph.

So for instance here:

![](https://user-images.githubusercontent.com/103693/41549636-c5d148e2-7326-11e8-9f8e-267c47ea6cec.png)

- The name is `"test"`;
- the tags are `[("hostname", string, "foo.local");  ("pid", int, 42)]` and  `[("hostname", string, "toto.com");  ("pid", int, 12)]`. They define families of graphs.
- the data forms the graphs  `y = MEM(t)` for both family of tags, where `MEM` is the projection of the collected data to values associated to the name `MEM`.  You could also plot `y = CPU(t)` (on a different graph) if you measure both memory and CPU usage on every data point.

Note: currently the filtering of tags can only be done on entry names for performance reason (as testing if a metric is enable or not is done in constant time). I am planning to add more complex queries at one point, but I don't want to have a too big runtime cost so need to careful.

The API provides way to build tags and data in a type-safe way. You define the `data` collection function and the shape of `tags` (like [here](https://github.com/samoht/irmin/blob/e49efa7f67ca34bb34fffce79250bac6f55d1e77/src/irmin/stats.ml#L31-L33)) when you define a new source with `Metrics.Src.v`. You pass the concrete `tags` when you create a new data-point with `Metrics.add src tags f`. If you want static tags, you should do that in your `tags` function like in [here](https://github.com/samoht/irmin/blob/e49efa7f67ca34bb34fffce79250bac6f55d1e77/src/irmin/stats.ml#L54).
